warehouse/accounts/views.py

```diff
@@ -735,6 +744,104 @@ def _get_two_factor_data(request, _redirect_to="/"):
     return two_factor_data
 
 
+@view_config(
+    route_name="accounts.verify-project-role",
+    uses_session=True,
+    permission="manage:user",
+    has_translations=True,
+)
+def verify_project_role(request):
+    token_service = request.find_service(ITokenService, name="email")
+    user_service = request.find_service(IUserService, context=None)
+
+    def _error(message):
+        request.session.flash(message, queue="error")
+        return HTTPSeeOther(request.route_path("manage.account"))
+
+    try:
+        token = request.params.get("token")
+        data = token_service.loads(token)
+    except TokenExpired:
+        return _error(request._("Expired token: request a new project role invite"))
+    except TokenInvalid:
+        return _error(request._("Invalid token: request a new project role invite"))
+    except TokenMissing:
+        return _error(request._("Invalid token: no token supplied"))
+
+    # Check whether this token is being used correctly
+    if data.get("action") != "email-project-role-verify":
+        return _error(request._("Invalid token: not an email verification token"))
+
+    user = user_service.get_user(data.get("user_id"))
+    if user != request.user:
+        return _error(request._("Invite is not valid"))
+
+    project = (
+        request.db.query(Project).filter(Project.id == data.get("project_id")).one()
+    )
```

Comment:
If we mounted this view at `/manage/project/{project_name}/...` instead, we could use URL traversal to get the current project, rather than having to do it manually in the view here.

We'd need to change how we generated the URL for the email slightly as well, but that shouldn't be an issue.

Reply:
Can't do this because it would require modifying the project ACL, so marking this as resolved.